utils/notifications.js

```diff
@@ -105,180 +110,184 @@ const sendEmail = async (emailTo, messageSubject, html, cc) => {
     });
 }
 
-const notificationHandler = async (message) => {
-    console.log("Message: " + message);
+async function notificationHandler(message) {
+  const scheduleAt = message.data ? Buffer.from(message.data, "base64").toString().trim() : null;
+  const notificationSpecArray = await getEmailNotifications(scheduleAt);
+  if (notificationSpecArray.length === 0) return;
+  const apiKey = await getSecrets();
+  sgMail.setApiKey(apiKey);
 
-    const publishedMessage = message.data ? Buffer.from(message.data, 'base64').toString().trim() : null;
-    const splitCharacters = '@#$'
-    let html = ``;
+  const notificationPromises = [];
+  for (const notificationSpec of notificationSpecArray) {
+    notificationPromises.push(handleNotificationSpec(notificationSpec));
+  }
 
-    if(!/@#\$/.test(publishedMessage)) {
-        const {PubSub} = require('@google-cloud/pubsub');
-        const pubSubClient = new PubSub();
-        const scheduleAt = publishedMessage;
+  await Promise.allSettled(notificationPromises);
+}
 
-        console.log(publishedMessage);
+async function handleNotificationSpec(notificationSpec) {
+  const primaryField = notificationSpec.primaryField;
+  let cutoffTime = new Date();
+  cutoffTime.setDate(cutoffTime.getDate() - notificationSpec.time.day);
+  cutoffTime.setHours(cutoffTime.getHours() - notificationSpec.time.hour);
+  cutoffTime.setMinutes(cutoffTime.getMinutes() - notificationSpec.time.minute);
+  
+  if (/\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{3}Z/.test(primaryField)) {
+    // Do nothing if primaryField is a timestamp and time isn't reached. Otherwise, take actions.
+    const scheduledTime = new Date(primaryField);
+    if (scheduledTime > cutoffTime) return;
+    await getParticipantsAndSendEmails({notificationSpec});
```

Comment:
suggestion: create object prior to first IF statement (i.e. specObject = {notificationSpec})

if you get into ELSE statement, add cutOffTimeStr and timeField to specObject

have one call to getParticipantsAndSendEmails(specObject)